### ISSUE TITLE
image mean and std preprocessing against (min, max) clip

### ISSUE THREAD

**Issue by machengcheng2016** [BorealisAI/advertorch#30]
Hello!
When I try to add *FGSM* perturbation on images in CIFAR-10 dataset, I've already done preprocessing as below, and the data range of images is no longer [0, 1].
`transform_test = transforms.Compose([
transforms.ToTensor(),
transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
])`
However, *FGSM* needs to be defined with 'data range' arguments, which are clip_min=0.0 and clip_max=1.0 by default. I do know these two arguments should be adjusted correspondingly, but how to compute this new data range after image preprocessing?
What comes to me is to create a toy image which only contains 0.0 and 1.0 pixels, and compute the data range after image preprocessing. But this seems a little clumsy. Is there any method in advertorch to handle this problem? As far as I know, there is a 'preprocessing=(mean, std)' together with 'bounds=(0, 1)' in *Foolbox* toolbox.

**Comment by gwding** [comment]
We usually treat the preprocessing as part of the model, so the input data range is always [0, 1] for images.
Have a look at 
https://github.com/BorealisAI/advertorch/blob/master/advertorch_examples/tutorial_attack_imagenet.ipynb and specifically the 2nd block to see if this is what you want.

```
from torchvision.models import resnet101
from advertorch.utils import predict_from_logits
from advertorch.utils import NormalizeByChannelMeanStd

normalize = NormalizeByChannelMeanStd(
    mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
model = resnet101(pretrained=True)
model.eval()
model = nn.Sequential(normalize, model)
model = model.to(device)
```

**Comment by machengcheng2016** [comment]
Thanks for your reply!
That's useful.

**Comment by RongDai430** [comment]
> Thanks for your reply!
> That's useful.

Hi! I met the same question with you, when trying to do FGSM on CIFAR-10 dataset, the data range after normalization is no longer [0,1], so if clamp it into [0,1], the acc drops.  After seeing **gwding**'s answer, I'm still confused about it. Could  you tell me how to deal with it more detaily without using any toolboxs ?

**Comment by gwding** [comment]
@RongDai430 The code snippet that I provide basically makes the data normalization part of the model, so clamping to [0, 1] is valid. Does this make sense?

**Comment by RongDai430** [comment]
> @RongDai430 The code snippet that I provide basically makes the data normalization part of the model, so clamping to [0, 1] is valid. Does this make sense?

Thanks for your reply!
After implementing by myself, now I totally understand it.
Thank you very much!
